### PR TITLE
Decouple agent builder from instrumentation

### PR DESCRIPTION
**Pull request by mcculls**
Begin moving classes that use `AgentBuilder` or its nested types to the new `agent-builder` module. Further PRs will update the `Instrumenter` design to allow different agent approaches and remove any remaining builder-related types.

Note: `DDCachingPoolStrategy` no longer directly implements `AgentBuilder.PoolStrategy` so we can share it between muzzle and the matching process regardless of the agent approach. Instead we use a small wrapper for `AgentBuilder` that maps between `AgentBuilder.PoolStrategy` and the shared concrete type.

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java [hunk 1]
```diff
@@ -2,38 +2,40 @@
...
-    return AgentInstaller.class.getClassLoader();
+    return Instrumenter.class.getClassLoader();
   }
 
   /** Return a classloader which can be used to look up bootstrap resources. */
   public static ClassLoader getBootstrapProxy() {
     if (getAgentClassLoader() instanceof DatadogClassLoader) {
       return ((DatadogClassLoader) getAgentClassLoader()).getBootstrapProxy();
     } else {
-      // in a unit test
```
Comment (bantonsson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java:33]: Maybe the comment is still relevant?

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java [hunk 2]
```diff
@@ -2,38 +2,40 @@
...
-
-  static {
-    try {
-      findLoadedClassMethod = ClassLoader.class.getDeclaredMethod("findLoadedClass", String.class);
-    } catch (final NoSuchMethodException | SecurityException e) {
-      throw new IllegalStateException(e);
-    }
+  /** Initialization-on-demand, only used during unit tests. */
+  static class BootstrapClassLoaderHolder {
+    static final ClassLoader unitTestBootstrapProxy = new BootstrapClassLoaderProxy();
```
Comment (bantonsson) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java:12]: Nice

**Review by bantonsson — APPROVED** [review]

**Review by devinsba — APPROVED** [review]
LGTM

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java (hunk 1 above)
Comment (mcculls) [reply dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/Utils.java:33]: 👍    f906a68c19

**Pull request merged**